### PR TITLE
feat(pm_auth): Added balance check for PM auth bank account

### PR DESCRIPTION
**Pull request by Sarthak1799**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

have added a balance check that would enable a check on the available amount in the selected bank account while doing a payment.
If the found balance is less than the intent amount, the payment would not go through.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Follow this for TCs - https://github.com/juspay/hyperswitch/pull/3047
In the payment create, use amount greater then 150 USD for the check to work

- Payment create - 
```
curl --location --request POST 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_0u355Qt5nbq8BhTZtWmbkwzQJ4JY77u0jOpjyIi3LRqIFziLTFKI8E6QwgFmaFc5' \
--data-raw '{
    "amount": 65400,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 65400,
    "customer_id": "StripeCustomer",
    "business_country": "US",
    "business_label": "default",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594430",
            "country_code": "+91"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}'
```
- Payment confirm 
```
curl --location --request POST 'http://localhost:8080/payments/pay_FKpvykZd46XL9grQwPlC/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: pk_dev_9cdfbdb024614790a4bf730b19a8c375' \
--data-raw '{
    "payment_token": "token_Yxogj2YxdeHjC316ZtM9",
    "client_secret": "pay_FKpvykZd46XL9grQwPlC_secret_KQLB47ADsiyrIhIpOvxv",
    "payment_method": "bank_debit",
    "payment_method_type": "ach"
}'
```
Response
```
{
    "error": {
        "type": "invalid_request",
        "message": "Payment amount is greater than the available amount in selected bank account",
        "code": "IR_06"
    }
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/pm_auth.rs [hunk 1]
```diff
@@ -716,6 +717,23 @@ pub async fn retrieve_payment_method_from_auth_service(
...
+        let required_conversion = util_types::FloatMajorUnitForConnector;
+        let converted_amount = required_conversion
+            .convert_back(balance, enums::Currency::USD)
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("Could not convert to MinorUnit")?;
+
+        if converted_amount < payment_intent.amount {
+            return Err((ApiErrorResponse::InvalidRequestData {
+                message:
+                    "Payment amount is less than the available amount in selected bank account"
```
Comment (Chethan-rao) [inline crates/router/src/core/pm_auth.rs:733]:
```suggestion
                    "Payment amount is greater than the available amount in selected bank account"
```

File: crates/router/src/core/pm_auth.rs [hunk 2]
```diff
@@ -716,6 +717,26 @@ pub async fn retrieve_payment_method_from_auth_service(
...
 
+    if let Some(balance) = bank_account.balance {
+        let required_conversion = util_types::FloatMajorUnitForConnector;
+        let converted_amount = required_conversion
+            .convert_back(
+                balance,
+                payment_intent.currency.unwrap_or(enums::Currency::USD),
+            )
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("Could not convert to MinorUnit")?;
```
Comment (Chethan-rao) [inline crates/router/src/core/pm_auth.rs:728]: could u change this error message to convey more on what conversion failed. Something like

```suggestion
            .attach_printable("Could not convert {something} to MinorUnit")?;
```

File: crates/router/src/core/pm_auth.rs [hunk 3]
```diff
@@ -716,6 +717,26 @@ pub async fn retrieve_payment_method_from_auth_service(
         .ok_or(ApiErrorResponse::InternalServerError)
         .attach_printable("Bank account details not found")?;
 
+    if let Some(balance) = bank_account.balance {
+        let required_conversion = util_types::FloatMajorUnitForConnector;
+        let converted_amount = required_conversion
+            .convert_back(
+                balance,
+                payment_intent.currency.unwrap_or(enums::Currency::USD),
```
Comment (Chethan-rao) [inline crates/router/src/core/pm_auth.rs:725]: Is it fine to make `USD` as fallback case here?

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

**Review by Chethan-rao — DISMISSED** [review]

File: crates/router/src/core/pm_auth.rs [hunk 4]
```diff
@@ -716,6 +717,23 @@ pub async fn retrieve_payment_method_from_auth_service(
...
 
+    if let (Some(balance), Some(currency)) = (bank_account.balance, payment_intent.currency) {
+        let required_conversion = util_types::FloatMajorUnitForConnector;
+        let converted_amount = required_conversion
+            .convert_back(balance, currency)
+            .change_context(ApiErrorResponse::InternalServerError)
+            .attach_printable("Could not convert FloatMajorUnit to MinorUnit")?;
+
+        if converted_amount < payment_intent.amount {
+            return Err((ApiErrorResponse::InvalidRequestData {
```
Comment (vspecky) [inline crates/router/src/core/pm_auth.rs:728]: This is more of a `PreconditionFailed` error. We're not reading any request fields.

**Review by vspecky — CHANGES_REQUESTED** [review]

File: crates/router/src/core/pm_auth.rs [hunk 5]
```diff
@@ -716,6 +717,23 @@ pub async fn retrieve_payment_method_from_auth_service(
...
+                message:
+                    "Payment amount is greater than the available amount in selected bank account"
```
Comment (vspecky) [inline crates/router/src/core/pm_auth.rs:730]: Also lets simplify the error message to something along the lines of `"... insufficient balance ..."`

**Review by vspecky — CHANGES_REQUESTED** [review]

**Review by vspecky — APPROVED** [review]

File: crates/router/src/core/pm_auth.rs [hunk 6]
```diff
@@ -66,6 +67,19 @@ pub async fn create_link_token(
 
     let pm_auth_key = format!("pm_auth_{}", payload.payment_id);
 
+    redis_conn
+        .exists::<Vec<u8>>(&pm_auth_key)
+        .await
+        .change_context(ApiErrorResponse::InvalidRequestData {
+            message: "Incorrect payment_id provided in request".to_string(),
+        })
+        .attach_printable("Corresponding pm_auth_key does not exist in redis")?
```
Comment (Chethan-rao) [inline crates/router/src/core/pm_auth.rs:76]: This should be `InternalServerError` as the first error propagation is due to redis error. If the bool is `false` then it should be 4xx.

Can be taken in the next PR.

**Pull request merged**